zoneinfo/src/parser.rs

```diff
@@ -0,0 +1,201 @@
+use core::{
+    iter::Peekable,
+    num::ParseIntError,
+    str::{Lines, SplitWhitespace},
+};
+
+use alloc::{borrow::ToOwned, string::String, vec, vec::Vec};
+
+use crate::{
+    rule::{Rule, RuleTable},
+    zone::ZoneTable,
+    ZoneInfoCompiler,
+};
+
+#[derive(Debug)]
+pub enum ZoneInfoParseError {
+    InvalidZoneHeader(u32),
+    MissingIdentifier(u32),
+    UnexpectedEndOfLine(u32, &'static str),
+    UnknownValue(u32, String),
+    ParseIntError(u32, ParseIntError, &'static str),
+}
+
+pub trait TryFromStr<C>: Sized {
+    type Error;
+    fn try_from_str(s: &str, context: &mut C) -> Result<Self, Self::Error>;
+}
+
+#[derive(Debug, Clone)]
+pub struct LineParseContext {
+    pub line_number: u32,
+    pub spans: Vec<&'static str>,
+}
+
+impl LineParseContext {
+    pub fn enter(&mut self, name: &'static str) {
+        self.spans.push(name);
+    }
+
+    pub fn span(&self) -> &'static str {
+        self.spans.last().expect("span not defined")
+    }
+
+    pub fn exit(&mut self) {
+        self.spans.pop();
+    }
+}
+
+impl Default for LineParseContext {
+    fn default() -> Self {
+        Self {
+            line_number: 1,
+            spans: vec!["undefined"],
+        }
+    }
+}
+
+pub trait ContextParse {
+    fn context_parse<T: TryFromStr<LineParseContext>>(
+        &self,
+        ctx: &mut LineParseContext,
+    ) -> Result<T, <T as TryFromStr<LineParseContext>>::Error>;
+}
+
+impl ContextParse for &str {
+    fn context_parse<T: TryFromStr<LineParseContext>>(
+        &self,
+        ctx: &mut LineParseContext,
+    ) -> Result<T, <T as TryFromStr<LineParseContext>>::Error> {
+        T::try_from_str(self, ctx)
+    }
+}
+
+impl ContextParse for String {
+    fn context_parse<T: TryFromStr<LineParseContext>>(
+        &self,
+        ctx: &mut LineParseContext,
+    ) -> Result<T, <T as TryFromStr<LineParseContext>>::Error> {
+        T::try_from_str(self, ctx)
+    }
+}
+
+impl TryFromStr<LineParseContext> for String {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, _: &mut LineParseContext) -> Result<Self, Self::Error> {
+        Ok(s.parse::<String>().expect("parse String is infallible"))
```

Comment:
question: Why use `s.parse::<String>().expect(...)` instead of just `String::from(s)`?

Reply:
Good question that I don't have an answer for other than I must have been super tired when I wrote that, to be honest.